// gs-utils/lib/fetch.js
/* Copyright © 2016 Ganchrow Scientific, SA all rights reserved */
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const utilities_1 = require("./utilities");
/**
 * Takes a regular key => val pair object and creates a simple Fetch class to return defaultValues if not exist
 * @class Fetch
 */
class Fetch {
    /**
     * Create a fetch instance
     * @param {FetchInterface} obj - key => val pair wrapped with fetch
     */
    constructor(obj) {
        this.map = new Map();
        if (utilities_1.isObject(obj.values)) {
            Object.keys(obj.values).forEach(k => {
                this.map.set(this.tryStringLower(k), obj.values[k]);
            });
        }
        this.defaultValue = obj.defaultValue;
    }
    fetch(k) {
        k = this.tryStringLower(k);
        if (!this.map.has(k)) {
            this.map.set(k, this.defaultValue);
        }
        return this.map.get(k);
    }
    tryStringLower(val) {
        return typeof val === 'string' ? val.toLowerCase() : val;
    }
}
exports.Fetch = Fetch;
