// gs-utils/lib/index.js
/* Copyright © 2015-2016 Ganchrow Scientific, SA all rights reserved */
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const utils = require("./utilities");
const gsLogger = require("./gsLogger");
const chunker_1 = require("./chunker");
const byteSizedChunker_1 = require("./byteSizedChunker");
const include_1 = require("./decorators/include");
const classify_1 = require("./decorators/classify");
const lowercaseArguments_1 = require("./decorators/lowercaseArguments");
const precondition_1 = require("./decorators/precondition");
const setOnce_1 = require("./decorators/setOnce");
const backoff_1 = require("./backoff");
exports.gsutils = Object.defineProperties(utils, {
    Chunker: { value: chunker_1.Chunker },
    ByteSizedChunker: { value: byteSizedChunker_1.ByteSizedChunker },
    decorators: { value: Object.defineProperties({}, {
            include: { value: include_1.include },
            classify: { value: classify_1.classify },
            lowercaseArguments: { value: lowercaseArguments_1.lowercaseArguments },
            setOnce: { value: setOnce_1.setOnce },
            precondition: { value: precondition_1.precondition }
        }) },
    Backoff: { value: backoff_1.Backoff },
    gsLogger: { value: gsLogger }
});
